string_manipulation.rb

```diff
@@ -1,21 +1,28 @@
 # A method to reverse a string in place.
 
 def string_reverse(my_string)
-  i = 0
-  j = my_string.length - 1
+  i = 0 # O(1) constant since variable is set to 0
+  j = my_string.length - 1 # 0(1) constant since j is set to the length of string - 1 (will always be the case regardless since it's fixed even if string is 1000 characters or 10, it will still be length -1)
 
-  while i < j
-    temp = my_string[i]
+  while i < j # while loop happens n amount of times depending on the length of my_string
+    temp = my_string[i] # O(1)
 
-    my_string[i] = my_string[j]
-    my_string[j] = temp
- 
-    i += 1
-    j -= 1
+    my_string[i] = my_string[j] # O(1)
+    my_string[j] = temp # O(1)
+
+    i += 1  # n times (operation: i + 1)
+    j -= 1  # n times (operation: j + 1)
   end
   return my_string
 end
 
+# Time complexity: n + n + n = 3n, constant can be eliminated as size of input grows
+# answer:  O(n)
+
+# Space complexity:  i and j are constant and memory is set aside for these variables.  Temp is variable that will be available after the function is run; my_string[i] and my_string[j] are just the indices in the array and memory has been allocated for my_string, which will typically be done for static array.
+
+#Hence even though, I want to say it's O(n), I think the space complexity is O(1)
```

Comment:
The space complexity will be O(1) because regardless of the length of the string passed in being of length 100 or 100,000, you'll only allocated storage for 3 variables i, j and temp. A new temp gets created for each iteration through the loop. Temp takes constant space (let's say 2 bytes). When the current iteration completes, those 2 bytes are given back to the system. When the next iteration begins, temp takes up 2 bytes again.